agent/hcp/telemetry_provider.go

```diff
@@ -0,0 +1,199 @@
+var _ telemetry.ConfigProvider = &hcpProviderImpl{}
+var _ telemetry.EndpointProvider = &hcpProviderImpl{}
+
+// hasher can be implemented to compute the hash of a dynamicConfig.
+type hasher interface {
+	hash(cfg *dynamicConfig) (uint64, error)
+}
+
+// hasherImpl uses the hashstructure library to compute dynamicConfig hash.
+type hasherImpl struct{}
+
+// hash returns a uint64 hash value for a dynamicConfig for equality comparisons.
+func (h *hasherImpl) hash(cfg *dynamicConfig) (uint64, error) {
+	return hashstructure.Hash(*cfg, hashstructure.FormatV2, nil)
+}
+
+// hcpProviderImpl holds telemetry configuration and settings for continuous fetch of new config from HCP.
+// it updates configuration, if changes are detected.
+type hcpProviderImpl struct {
+	// cfg holds configuration that can be dynamically updated.
+	cfg *dynamicConfig
+
+	// A reader-writer mutex is used as the provider is read heavy.
+	// OTEL components access telemetryConfig during metrics collection and export (read).
+	// Meanwhile, config is only updated when there are changes (write).
+	rw sync.RWMutex
+	// hcpClient is an authenticated client used to make HTTP requests to HCP.
+	hcpClient client.Client
+	// ticker is a reference to the time ticker that can be reset when refreshInterval changes.
+	ticker *time.Ticker
+	hasher hasher
+}
+
+// dynamicConfig is a set of configurable settings for metrics collection, processing and export.
+// fields MUST be exported to compute hash for equals method.
+type dynamicConfig struct {
+	Endpoint *url.URL
+	Labels   map[string]string
+	Filters  *regexp.Regexp
+	// refreshInterval controls the interval at which configuration is fetched from HCP to refresh config.
+	RefreshInterval time.Duration
+}
+
+// NewHCPProvider initializes and starts a HCP Telemetry provider with provided params.
+func NewHCPProvider(ctx context.Context, hcpClient client.Client, telemetryCfg *client.TelemetryConfig) (*hcpProviderImpl, error) {
+	refreshInterval := telemetryCfg.RefreshConfig.RefreshInterval
+	if refreshInterval <= 0 {
+		return nil, fmt.Errorf("invalid refresh interval: %d", refreshInterval)
+	}
+
+	cfg := &dynamicConfig{
+		Endpoint:        telemetryCfg.MetricsConfig.Endpoint,
+		Labels:          telemetryCfg.MetricsConfig.Labels,
+		Filters:         telemetryCfg.MetricsConfig.Filters,
+		RefreshInterval: refreshInterval,
+	}
+
+	t := new(hcpClient, cfg)
+
+	go t.run(ctx, t.ticker.C)
+
+	return t, nil
+}
+
+func new(hcpClient client.Client, cfg *dynamicConfig) *hcpProviderImpl {
+	return &hcpProviderImpl{
+		cfg:       cfg,
+		hcpClient: hcpClient,
+		hasher:    &hasherImpl{},
+		ticker:    time.NewTicker(cfg.RefreshInterval),
+	}
+}
+
+// run continously checks for updates to the telemetry configuration by making a request to HCP.
+// Modification of config only occurs if changes are detected to decrease write locks that block read locks.
+func (t *hcpProviderImpl) run(ctx context.Context, tick <-chan time.Time) {
```

Comment:
Remove the line 96 and put into 102. We don't need to test the ticker functionality as it is a library.